src/lib/reasoners/rel_utils.ml

```diff
@@ -197,182 +197,210 @@ module type Constraint = sig
+  val add : ?pending:bool -> ex:Explanation.t -> Constraint.t -> t -> t
+  (** [add ~ex c t] adds the constraint [c] to the set [t].
 
-      Use this to ensure that the representation is always normalized.
+      The explanation [ex] justifies that the constraint [c] holds.
+
+      The constraint is only added to the pending set if it was not already
+      active (i.e. previously added). Setting the [pending] optional argument to
+      [true] forces the constraint to be marked as pending even if it is already
+      active. *)
+
+  val subst : ex:Explanation.t -> X.r -> X.r -> t -> t
+  (** [subst ~ex p v t] replaces all instances of [p] with [v] in the
+      constraints.
 
       The explanation [ex] justifies the equality [p = v]. *)
 
-  val add : t -> Constraint.t -> t
-  (** [add c cs] adds the constraint [c] to [cs]. *)
+  val notify : X.r -> t -> t
+  (** [notify r t] marks all constraints involving [r] (i.e. all constraints
+      that have [r] as one of their arguments) as pending.
 
-  val propagate_fresh :  t -> Constraint.domain -> t * Constraint.domain
-  (** [propagate_fresh cs acc] propagates the fresh constraints and returns the
-      new domain, as well as a copy of the constraint set with no fresh
-      constraints.
+      This function should be used when the domain of [r] is updated, if
+      domains are tracked for all representatives. *)
+
+  val notify_leaf : X.r -> t -> t
+  (** [notify_leaf r t] marks all constraints that have [r] as a leaf (i.e.
+      all constraints that have at least one argument [a] such that [r] is in
+      [X.leaves a]) as pending.
 
-      Fresh constraints are constraints that were never propagated yet. *)
+      This function should be used when the domain of [r] is updated, if
+      domains are tracked for leaves only. *)
 
-  val fold_r : (X.r -> 'a -> 'a) -> t -> 'a -> 'a
-  (** [fold_r f cs acc] folds [f] over any representative [r] that is currently
-      associated with a constraint (i.e. at least one constraint currently
-      applies to [r]). *)
+  val fold_args : (X.r -> 'a -> 'a) -> t -> 'a -> 'a
+  (** [fold_args f t acc] folds [f] over all the term representatives that are
+      arguments of at least one constraint. *)
+
+  val next_pending : t -> Constraint.t explained * t
+  (** [next_pending t] returns a pair [c, t'] where [c] was pending in [t] and
+      [t'] is identical to [t], except that [c] is no longer a pending
+      constraint.
 
-  val propagate : t -> X.r -> Constraint.domain -> Constraint.domain
-  (** [propagate cs r dom] propagates the constraints associated with [r] in the
-      constraint set [cs] and returns the new domain map after propagation. *)
+      @raise Not_found if there are no pending constraints. *)
 end = struct
-  module IM = Util.MI
   module MX = Shostak.MXH
 
-  module CS = Set.Make(Constraint)
+  module CS = Set.Make(struct
+      type t = Constraint.t explained
+
+      let compare a b = Constraint.compare a.value b.value
```

Comment:
I don't know if it was like this before but the fact we ignore explanations here means we cannot store several times the same constraint with not comparable explanations (by not comparable explanations, I mean there is no one of them which is subsumed by the other one).

We should document this point.